studio-docs/source/metrics/operation-signatures.mdx

```diff
@@ -0,0 +1,133 @@
+
+<img src="../img/separate-operations.jpg" alt="Separate operations with same name in Studio" class="screenshot" width="500" />
+
+To help Studio identify functionally identical operations, Apollo Server generates the **signature** for each operation it reports to Studio. This signature is a normalized representation for an operation with deterministic field order, whitespace, and values for in-line argument values.
+
+## Why do we need an operation signature?
+
+Consider the following operations:
+
+```graphql{10-11,17}
+query GetPostDetails($postId: String!) {
+  post(id: $postId) {
+    author
+    content
+  }
+}
+
+query GetPostDetails($postId: String!) {
+  post(id: $postId) {
+    content # Different field order
+    author
+  }
+}
+
+query GetPostDetails($postId: String!) {
+  post(id: $postId) {
+    writer: author # Field alias
+    content
+  }
+}
+```
+
+Despite some cosmetic differences, all of these operations execute identically on a particular GraphQL server. Therefore, Studio should group them when displaying performance information.
+
+Helpfully, Apollo Server's [signature algorithm](#signature-algorithm) generates the _exact same operation signature_ for all of these operations, which means Studio _does_ group them.
+
+## Signature algorithm
+
+[Jump to an example](#example-signature)
+
+Apollo Server's current signature algorithm performs the following modifications on an operation to generate its signature:
+
+### 1. Transform in-line argument values
+
+If an operation includes any in-line argument values, those values are transformed according to their type:
+
+* `Boolean` and `enum` values are _preserved_.
+* `Int` and `Float` values are replaced with `0`.
+* `String`, list, and object values are replaced with their "empty" counterpart (`""`, `[]`, or `{}`).
+
+> Argument values provided as GraphQL variable names are preserved.
+
+### 2. Remove extraneous characters
+
+All of an operation's [ignored tokens](http://spec.graphql.org/October2021/#sec-Language.Source-Text.Ignored-Tokens) (including comments and redundant whitespace) are removed. Single spaces are preserved only when required for parsing the request.
+
+If the operation document includes _multiple_ operations, then operations besides the _executed_ operation are removed.
+
+If the operation document includes fragments that _aren't used_ by the executed operation, then those fragments are removed (other fragments are preserved).
+
+### 3. Reorder definitions
+
+The definition of the executed operation appears _first_, followed by any applicable fragment definitions.
+
+#### Fragment definitions
+
+If any fragment definitions contain _sub_-fragments, _all_ top-level fragment definitions appear before _all_ second-level fragment definitions, and so on.
```

Comment:
I might have interpreted the old version incorrectly here:

> Fragment definitions appear in order of first reachability when traversing spread-first, depth-second.

I couldn't tell if that meant traversing the various object scopes of the query, or traversing the fragments themselves.

Reply:
I don't know why it said that but it's not correct (at least not these days, maybe it once was?).

Our sortAST does this

```
        // Use sortBy instead of sorted because 'definitions' is not optional.
        // The sort on "kind" places fragments before operations within the document
        definitions: sortBy(node.definitions, 'kind', 'name.value'),
```

So fragments come before the operation and fragments are sorted by name.